strings/base_com_ptr.h

```diff
@@ -59,33 +59,31 @@ WINRT_EXPORT namespace winrt
 
         com_ptr& operator=(com_ptr const& other) noexcept
         {
-            copy_ref(other.m_ptr);
+            com_ptr cpy(other);
+            swap(*this, cpy);
             return*this;
         }
 
         com_ptr& operator=(com_ptr&& other) noexcept
         {
-            if (this != &other)
-            {
-                release_ref();
-                m_ptr = std::exchange(other.m_ptr, {});
-            }
-
+            com_ptr cpy(std::move(other));
```

Comment:
This means that `x = std::move(x)` temporarily leaves x null, which could be surprising. While it's unlikely that somebody will write exactly that, they may end up doing so through a series of wrappers, like a `std::tuple::get` method that returns an rvalue reference.

Reply:
You are right. Do you have a suggestion for a good solution?
The only thing that comes to my mind is to check again for self-assignment.

Reply:
Take by value and you don't need to check for self-assignment. This is one of the motivation for the copy and swap idiom :D

Reply:
@roxk: Semantically that shouldn't change anything (now the user code would have to check for self assignment if it wants to avoid the transitional null state).
@oldnewthing :  What exactly would be the problem with x bein temporarily null? How would external code observe this?

Reply:
> Yes, this is technically UB, but it nevertheless happens, and the "temporarily null" behavior is unexpected.

This might be a generational thing (I only started to work with c++ in ernest around the time c++11 gotfinalized), but to be honest: I don't expect anything sane to happen if I violate the c++ language rules with regards to multithreading. 

Anyway: If never null even under violation of c++ multithreading rules is a desired property, then making this a by-value-parameter would definetly be out of the question, as you then would do a unconditional move. You have to check for self assignment instead as suggested by @afrankeser.


Btw.: Your example is using copy assignement, not move assignment.

Reply:
> Take by value and you don't need to check for self-assignment. This is one of the motivation for the copy and swap idiom :D
Yes I know but one of the motivation of the PR is to avoid temporary nullptr during assignment. I will give it a try and readd the check.

Reply:
I don't see any problems with the PR as is (but that isn't meant to be a proper review)